packages/hadron-document/test/document.test.ts

```diff
@@ -1558,65 +1558,373 @@ describe('Document', function () {
+                            field: { $literal: '$a.b' },
+                            input: '$$ROOT',
+                          },
+                        },
+                      },
+                      in: {
+                        $concatArrays: [
+                          { $slice: ['$$input', 0, 3] },
+                          [{ $literal: new Int32(3) }],
+                          { $slice: ['$$input', 4, 2147483647] },
+                        ],
+                      },
+                    },
+                  },
+                },
+              },
+            },
+          ],
+        });
       });
-      doc.get('a')?.get(2)?.remove();
-      expect(doc.generateUpdateUnlessChangedInBackgroundQuery()).to.deep.equal({
-        query: {
-          _id: null,
-          a: [new Int32(1), new Int32(2), new Int32(3)],
-        },
-        updateDoc: {
-          $set: { a: [new Int32(1), new Int32(2)] },
-        },
+      it('can remove array elements in the middle of the array', function () {
+        const doc = new Document({
+          '$a.b': [1, 2, 3],
+        });
+        doc.get('$a.b')?.get(1)?.remove();
+        expect(
+          doc.generateUpdateUnlessChangedInBackgroundQuery()
+        ).to.deep.equal({
+          query: {
+            _id: null,
+            $expr: {
+              $eq: [
+                {
+                  $getField: { field: { $literal: '$a.b' }, input: '$$ROOT' },
+                },
+                {
+                  $literal: [new Int32(1), new Int32(2), new Int32(3)],
+                },
+              ],
+            },
+          },
+          updateDoc: [
+            {
+              $replaceWith: {
+                $setField: {
+                  field: { $literal: '$a.b' },
+                  input: '$$ROOT',
+                  value: { $literal: [new Int32(1), new Int32(3)] },
+                },
+              },
+            },
+          ],
+        });
+      });
+      it('can remove array elements at the end of the array', function () {
+        const doc = new Document({
+          '$a.b': [1, 2, 3],
+        });
+        doc.get('$a.b')?.get(2)?.remove();
+        expect(
+          doc.generateUpdateUnlessChangedInBackgroundQuery()
+        ).to.deep.equal({
+          query: {
+            _id: null,
+            $expr: {
+              $eq: [
+                {
+                  $getField: { field: { $literal: '$a.b' }, input: '$$ROOT' },
+                },
+                {
+                  $literal: [new Int32(1), new Int32(2), new Int32(3)],
+                },
+              ],
+            },
+          },
+          updateDoc: [
+            {
+              $replaceWith: {
+                $setField: {
+                  field: { $literal: '$a.b' },
+                  input: '$$ROOT',
+                  value: { $literal: [new Int32(1), new Int32(2)] },
+                },
+              },
+            },
+          ],
+        });
```

Comment:
I don’t necessarily mind having these fairly verbose/repetitive test object assertions, but if we feel that this is too noisy, I’m happy to take suggestions on how to improve that. I guess ideally some kind of snapshot testing where it’s easier to do bulk updates to expected results would be a good pick, but I don’t want to introduce entirely new paradigms here.